java-checks/src/main/java/org/sonar/java/checks/tests/TooManyAssertionsCheck.java

```diff
@@ -0,0 +1,123 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.tests;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.sonar.check.Rule;
+import org.sonar.check.RuleProperty;
+import org.sonar.java.model.ModifiersUtils;
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodReferenceTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Modifier;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import static org.sonar.java.checks.helpers.UnitTestUtils.ASSERTION_INVOCATION_MATCHERS;
+import static org.sonar.java.checks.helpers.UnitTestUtils.ASSERTION_METHODS_PATTERN;
+import static org.sonar.java.checks.helpers.UnitTestUtils.REACTIVE_X_TEST_METHODS;
+import static org.sonar.java.checks.helpers.UnitTestUtils.TEST_METHODS_PATTERN;
+import static org.sonar.java.checks.helpers.UnitTestUtils.isUnitTest;
+import static org.sonar.java.model.ExpressionUtils.methodName;
+
+@Rule(key = "S5961")
+public class TooManyAssertionsCheck extends BaseTreeVisitor implements JavaFileScanner {
+
+  private static final int DEFAULT_MAX = 10;
+
+  @RuleProperty(key = "MaximumAssertionNumber", description = "The maximum authorized number of assertions in a test method", defaultValue = "" + DEFAULT_MAX)
+  public int maximumAssertionNumber = DEFAULT_MAX;
+
+  private final Map<Symbol, Integer> assertionsInMethod = new HashMap<>();
+  private JavaFileScannerContext context;
+
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    this.context = context;
+    assertionsInMethod.clear();
+    scan(context.getTree());
+    assertionsInMethod.clear();
+  }
+
+  @Override
+  public void visitMethod(MethodTree methodTree) {
+    if (ModifiersUtils.hasModifier(methodTree.modifiers(), Modifier.ABSTRACT)) {
+      return;
+    }
+
+    if (isUnitTest(methodTree) && numberOfAssertions(methodTree.symbol()) > maximumAssertionNumber) {
+      context.reportIssue(this, methodTree.simpleName(), String.format("Refactor this method in order to have less than %d assertions.", maximumAssertionNumber));
+    }
+  }
+
+  private int numberOfAssertions(Symbol symbol) {
+    if (!assertionsInMethod.containsKey(symbol)) {
+      assertionsInMethod.put(symbol, 0);
+      Tree declaration = symbol.declaration();
```

Comment:
you already have the declaration in the `visitMethod()` method (L70), why not directly use it instead of the symbol?
So you would not have to jump into the declaration. You can these the symbol inside this method the.

Reply:
I think we need declaration in case we want to detect a local method which contains assertions. 
`  @Test
  void test3() { // Noncompliant [[sc=8;ec=13]]{{Refactor this method in order to have less than 2 assertions.}}
    assertEquals(2, f(2));
    assertEquals(3, f(1));
    customAssert();
  }

  void customAssert() {
    assertEquals(2, f(2));
    assertEquals(3, f(1));
  }
`

In this case `tree` points to invocation of `customAssert()` while to understand whether this is an assert itself we need its declaration.